driver: bluetooth: hci: Add NXP BT module support on RT1170 EVKB

lylezhu2012 opened the pull request:
Implement UART firmware download driver for NXP BT module.

Only support Murata 2EL M.2 Module on RT1170 EVKB.

msmttchr commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4108,6 +4108,14 @@ int bt_enable(bt_ready_cb_t cb)
 	k_thread_name_set(&bt_workq.thread, "BT RX");
 #endif
 
```
@lylezhu2012, I noticed that this `init` function is called just before the `open`, wouldn’t be possible to perform `init` actions within `open` ?

lylezhu2012 replied:
updated

dleach02 commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,1178 @@
+
+	uart_irq_rx_enable(uart_dev);
+
+	err = fw_uploading(bt_fw_bin, bt_fw_bin_len);
```
@lylezhu2012, We need to put together a request to the project to handle the binary blob for this firmware. I'll take this offline with you to get this setup.

lylezhu2012 replied:
This PR is used to implement a driver to download controller firmware via UART.
I have a guide here https://github.com/zephyrproject-rtos/zephyr/pull/70815/commits/7f5f360ec1060f5eba55d3d009ee7fa3ebb76a10#diff-dccbb25cc34086042193126bd5b09e2b1ff09d3f24fb18da73095b6c2244560cR252-R275 to show how to download firmware.

Once the binary blob is done, I will update the driver and guide.

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -502,13 +502,35 @@ int __weak bt_hci_transport_setup(const struct device *dev)
 
 	LOG_DBG("");
 
+#if defined(CONFIG_BT_HCI_INIT)
+	ret = h4_init();
+	if (ret < 0) {
+		return ret;
+	}
+#endif
```
Couldn't you just implement your own `bt_hci_transport_setup()` instead of adding yet another vendor "setup" function?

lylezhu2012 replied:
Yes. I updated the PR.

jhedberg commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,1178 @@
+	}
+}
+
+int bt_nxp_ctlr_init(void)
```
Seems like this should be static?

lylezhu2012 replied:
Updated.

jhedberg commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,1178 @@
+#define A6REQ_PAYLOAD_LEN 8U
+#define AbREQ_PAYLOAD_LEN 3U
+
+#define CRC_ERR_BIT          1 << 0U
+#define NAK_REC_BIT          1 << 1U
+#define TIMEOUT_REC_ACK_BIT  1 << 2U
+#define TIMEOUT_REC_HEAD_BIT 1 << 3U
+#define TIMEOUT_REC_DATA_BIT 1 << 4U
+#define INVALID_CMD_REC_BIT  1 << 5U
+#define WIFI_MIC_FAIL_BIT    1 << 6U
+#define BT_MIC_FAIL_BIT      1 << 7U
```
these could probably use the `BIT()` macro instead

lylezhu2012 replied:
Updated.

jhedberg commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,1178 @@
+
+	speed = DT_PROP(DT_CHOSEN(zephyr_bt_uart), current_speed);
+	speed = DT_PROP_OR(HCI_UART_CHILD, hci_operation_speed, speed);
+	uart_dev_data.primary_speed = DT_PROP_OR(HCI_UART_CHILD, fw_download_primary_speed, speed);
+	uart_dev_data.secondary_speed =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_secondary_speed, speed);
+
+	flowcontrol_of_hci = DT_PROP_OR(HCI_UART_CHILD, hw_flow_control, false) ? true : false;
+	uart_dev_data.primary_flowcontrol =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_primary_flowcontrol, false) ? true : false;
+	uart_dev_data.secondary_flowcontrol =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_secondary_flowcontrol, false) ? true : false;
```
The ternary operator seems redundant here, i.e. `DT_PROP_OR()` should in itself already evaluate to a boolean value.

lylezhu2012 replied:
If the property exists, the value is `1`. It needs to be cast to true. Cast it using `bool`.

jhedberg reviewed (changes requested):
Looks pretty good. Just a few minor comments.

msmttchr commented on dts/bindings/bluetooth/nxp,bt-hci-uart.yaml:
```diff
@@ -0,0 +1,50 @@
+  fw-download-primary-speed:
+    type: int
+    description: |
+      HCI UART primary boudrate for FW download operation.
+      If not defined bus/current-speed will be used as default.
```
```suggestion
      HCI UART primary baudrate for FW download operation.
      If not defined bus/current-speed will be used as default.
```

lylezhu2012 replied:
updated

msmttchr commented on dts/bindings/bluetooth/nxp,bt-hci-uart.yaml:
```diff
@@ -0,0 +1,50 @@
+  fw-download-primary-flowcontrol:
+    type: boolean
+    description: |
+      Flowcontrol setting for primary peed.
```
```suggestion
      Flow control setting for primary speed.
```

lylezhu2012 replied:
Fixed

msmttchr commented on dts/bindings/bluetooth/nxp,bt-hci-uart.yaml:
```diff
@@ -0,0 +1,50 @@
+  fw-download-secondary-speed:
+    type: int
+    description: |
+      HCI UART secondary boudrate for FW download operation.
```
```suggestion
      HCI UART secondary baudrate for FW download operation.
```

lylezhu2012 replied:
Fixed

msmttchr commented on dts/bindings/bluetooth/nxp,bt-hci-uart.yaml:
```diff
@@ -0,0 +1,50 @@
+  fw-download-secondary-flowcontrol:
+    type: boolean
+    description: |
+      Flowcontrol setting for secondary peed.
```
```suggestion
      Flow control setting for secondary speed.
```

lylezhu2012 replied:
fixed

msmttchr commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -0,0 +1,1178 @@
+#define HCI_UART_CHILD_PREFIX(x) x
+#define HCI_UART_CHILD           DT_FOREACH_CHILD(DT_CHOSEN(zephyr_bt_uart), HCI_UART_CHILD_PREFIX)
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios)
+struct gpio_dt_spec sido_reset = GPIO_DT_SPEC_GET(HCI_UART_CHILD, sdio_reset_gpios);
```
```suggestion
struct gpio_dt_spec sdio_reset = GPIO_DT_SPEC_GET(HCI_UART_CHILD, sdio_reset_gpios);
```
Is this a misspelling? If yes also code below needs to be fixed

lylezhu2012 replied:
fixed

msmttchr reviewed (commented):
Few minor likely typo reported.

msmttchr reviewed (approved):
Ok for me. Thanks.

dleach02 commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -234,3 +234,45 @@ config BT_SILABS_HCI_BUFFER_MEMORY
+	  of Murata 2EL M.2 Moudle. It could be downloaded here
+	  https://github.com/NXP/wifi_nxp/blob/MCUX_2.15.000/wifi_bt_firmware/nw61x/uart_nw61x_se.h
+	  And it needs to be placed into the folder <zephyrproject>/zephyr/
+	  drivers/bluetooth/hci.
```
Let us discuss this internally before approving for merging.

dleach02 reviewed (dismissed):
I want to discuss this internally before removing the block

dleach02 commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -234,3 +234,45 @@ config BT_SILABS_HCI_BUFFER_MEMORY
+	  of Murata 2EL M.2 Moudle. It could be downloaded here
+	  https://github.com/NXP/wifi_nxp/blob/MCUX_2.15.000/wifi_bt_firmware/nw61x/uart_nw61x_se.h
+	  And it needs to be placed into the folder <zephyrproject>/zephyr/
+	  drivers/bluetooth/hci.
```
This needs to be driving via binary blob support in the NXP HAL.

lylezhu2012 replied:
Yes. Updated.

dleach02 replied:
@lylezhu2012, to update you will need to change the west.yml to point to the NXP HAL PR: https://github.com/zephyrproject-rtos/hal_nxp/pull/381

lylezhu2012 replied:
Yes. It is updated here https://github.com/zephyrproject-rtos/zephyr/pull/70815/commits/a5c46c9e4bb2f3cee1c977b4878fabc5ab3067ad

lylezhu2012 committed:
Samples: bluetooth: hf: Add board specific configuration

Overlay sram when BT is enabled on board MIMXRT1170 EVKB.

Select NXP NW612 chipset.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

mmahadevan108 commented on boards/nxp/mimxrt1170_evk/mimxrt1170_evk-pinctrl.dtsi:
```diff
@@ -229,6 +229,17 @@
 		};
 	};
 
+	pinmux_lpuart2_flowcontrol: pinmux_lpuart2_flowcontrol {
```
Please update the documentation and maybe add a `Bluetooth` section and indicate the command to fetch the blobs.

lylezhu2012 replied:
Updated.

lylezhu2012 committed:
boards: mimxrt1170_evk: doc: add blobs support

Add `fetch Binary Blobs` section to guide how to
fetch binary blobs.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

pdgendt commented on modules/hal_nxp/bt_controller/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+    message(FATAL_ERROR "Unsupported controller. Please select a BT conntroller, refer to ./driver/bluetooth/hci/Kconfig.nxp")
+endif()
+
+if(EXISTS ${blob_file})
```
You can verify the blob with

```cmake
zephyr_blobs_verify(FILES ${blob_file} REQUIRED)
```

This will fail if the file is missing, or the SHA256 doesn't match.

dleach02 replied:
I didn't know this one

lylezhu2012 replied:
updated.
One problem is that zephyr_blobs_verify will report the same error if the blobs file does not exist or the blobs file is invalid. Errors cannot be intuitively known to the user as to the cause of the error.
Therefore, this file existence check is retained. If the file does not exist, prompt the user to fetch blobs through a message.
zephyr_blobs_verify is mainly used to check whether there are SHA256 errors.

pdgendt replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/73051

lylezhu2012 committed:
module: hal_nxp: bt_controller: Verify if Blob is valid

Use zephyr_blobs_verify to check the blob file is valid
or not.

The function zephyr_blobs_verify will check if the file
exists. And it checks if the file is valid.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Samples: bluetooth: hf_ag: Add board specific configuration

Overlay sram when BT is enabled on board MIMXRT1170 EVKB.

Select NXP NW612 chipset for Handsfree_ag example

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

dleach02 commented:
@lylezhu2012 we merged the HAL and went ahead and updated your PR with the new SHA

pdgendt commented on drivers/bluetooth/hci/Kconfig.nxp:
```diff
@@ -0,0 +1,18 @@
+
+config BT_NXP_NW612
+	bool "NXP IW612 Chipset"
+	help
+		NXP IW612 Chipset supports Wi-Fi® 802.11a/b/g/n/ac/ax + Bluetooth® 5.3
+		BR/EDR/LE + IEEE802.1.5.4 up to 601 Mbps data rate on Wi-Fi® and 2Mbps
+		data rate on Bluetooth®. 4-wire UART@3M baud is supported. PCM for
+		audio is also supported.
+		Details of the module could be fond on https://www.nxp.com/products/
+		wireless-connectivity/wi-fi-plus-bluetooth-plus-802-15-4/2-4-5-ghz-
+		dual-band-1x1-wi-fi-6-802-11ax-plus-bluetooth-5-4-plus-802-15-4-tri-
+		radio-solution:IW612.
```
Nit: indent help with a tab and 2 spaces

lylezhu2012 replied:
Fixed.

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+static unsigned long crc_table[256U];
+static bool made_table;
+
+static void fw_upload_initCrc8(void)
```
Avoid using camelCasing, also can't you use the CRC functions from `zephyr/sys/crc.h`?

lylezhu2012 replied:
Fixed.

CRC8 could be replaced by zephyr/sys/crc.h. 
But CRC32 can not be replaced.

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+	int64_t end;
+	char c;
+
+	end = k_uptime_get() + CONFIG_BT_NXP_CTLR_WAIT_HDR_SIG_TIMEOUT;
```
Use `k_timepoint_t`

lylezhu2012 replied:
It seems `k_timepoint_t` depends on CONFIG_SYS_CLOCK_EXISTS. If CONFIG_SYS_CLOCK_EXISTS cannot be enabled, the feature cannot be supported.
It is a limitation for this PR.

For k_uptime_get(), there is not such limitation.

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+		if (crc != buffer[AbREQ_PAYLOAD_LEN + REQ_HEADER_LEN]) {
+			return -EINVAL;
+		}
+	} else {
+	}
```
Empty block?

lylezhu2012 replied:
Added a LOG_ERR message.

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+	}
+}
+
+static int fw_upload_checkReqCrc(uint8_t *buffer, uint8_t uiReq)
```
Same, use snake_case

lylezhu2012 replied:
Updated

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+	return 0;
+}
+
+#define FW_UPLOAD_CHANGE_TIMEOUT_RETRY_COUNT 6
```
Move defines to the top of the file

lylezhu2012 replied:
done

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+					retry--;
+					fw_upload_send_ack(V3_TIMEOUT_ACK);
+				} else {
+					return -ETIME;
```
A timeout error typically uses `-EAGAIN`.

lylezhu2012 replied:
Updated

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+
+	while (!done) {
+		err = fw_upload_read_data(&fw_upload.hdr_sig, 1);
+		if (err >= 0) {
```
Check for negative values, otherwise you might have an infinite loop.

lylezhu2012 replied:
Updated

pdgendt commented on drivers/bluetooth/hci/hci_nxp_setup.c:
```diff
@@ -0,0 +1,1178 @@
+	}
+}
+
+static int fw_upload_write_hdr_and_data(uint16_t len_to_send, uint8_t *buffer, bool new_speed)
```
A general remark, but I'll take this function as an example;

The number of indentations and possible flows, makes it hard to review IMO. Maybe the process is just that complex, but reading the code doesn't make it easier.

lylezhu2012 replied:
Change `fw_upload_write_hdr_and_data` to `fw_upload_write_hdr_and_payload`.

For the function name of `drivers/bluetooth/hci/hci_nxp_setup.c`, it is determined by the firmware download protocol of the controller. The naming is more consistent with the protocol.

pdgendt replied:
I didn't mean the name of the function actually, but rather the body...

lylezhu2012 commented:
Due to the conflict, rebased the latest. And the updated version of hal_nxp has been synced, remove change of version update of hal_nxp in west.yml.

lylezhu2012 commented:
Due to the conflict, rebased the latest. And the updated version of hal_nxp has been synced, remove change of version update of hal_nxp in west.yml.

pdgendt commented on samples/bluetooth/handsfree_ag/boards/mimxrt1170_evk_mimxrt1176_cm7_B.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_BT_NXP_CTLR=y
+#select NXP NW612 Chipset
+CONFIG_BT_NXP_NW612=y
+#select CRC, it is depended by CONFIG_BT_NXP_CTLR
+CONFIG_CRC=y
```
Why not select it there then?

lylezhu2012 replied:
Current, `BT_NXP_CTLR` use 'depend on' to require `CRC`. 

It is changed from "depend on" to "select".

lylezhu2012 commented:
@dleach02 , @jhedberg , @pdgendt , @mmahadevan108 , Could you please take a look at it.

dleach02 commented:
@pdgendt can you confirm @lylezhu2012 response to your feedback

@jhedberg can you provide a review

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -237,3 +237,31 @@ config BT_SILABS_HCI_BUFFER_MEMORY
+	depends on BT_H4
+	select CRC
+	help
+	  Enables support for NXP Bluetooth Controller.
+	  More inforamtion about NXP Bluetooth profuct could be found on
+	  https://www.nxp.com/products/wireless-connectivity/wi-fi-plus-bluetooth-plus-802-15-4:WIFI-BLUETOOTH
```
Should this perhaps mention that this is an extension of the H:4 driver, rather than being a complete module in itself? It's kind of implicit through the `depends on BT_H4` but I think it'd be good to explicitly mention it, to not give the impression that this option provides a complete HCI driver.

In fact, considering that it's an extension of the H4 driver, would something like `BT_H4_NXP_SETUP` or `BT_H4_NXP_CTLR` perhaps make more sense as the option name?

lylezhu2012 replied:
Updated

jhedberg commented on boards/nxp/mimxrt1170_evk/mimxrt1170_evk_mimxrt1176_cm7_B.overlay:
```diff
@@ -79,3 +80,25 @@
 	};
 };
+
+m2_hci_uart: &lpuart2 {
+	pinctrl-0 = <&pinmux_lpuart2_flowcontrol>;
+	pinctrl-1 = <&pinmux_lpuart2_sleep>;
+	pinctrl-names = "default", "sleep";
+
+	m2_bt_module {
+		compatible = "nxp,bt-hci-uart";
+		sdio-reset-gpios = <&gpio9 15 GPIO_ACTIVE_HIGH>;
+		w-disable-gpios = <&gpio9 30 GPIO_ACTIVE_HIGH>;
+		hci-operation-speed = <115200>;
+		hw-flow-control;
+		fw-download-primary-speed = <115200>;
+		fw-download-secondary-speed = <3000000>;
+		fw-download-secondary-flowcontrol;
+	};
+};
```
Just a note here, as part of #72323 we'll probably need to do the same as for the `cyw43xxx.c` driver (see https://github.com/zephyrproject-rtos/zephyr/pull/72323#issuecomment-2147417411), i.e. place a `zephyr,bt-hci-uart` node between the UART node and the `nxp,bt-hci-uart` node (i.e. the UART node would become the grandparent rather than the parent of this driver, which requires some small code changes).

lylezhu2012 committed:
driver: bluetooth: hci: Support NXP BT CTLR FW blobs

Add Kconfig.nxp to support NXP Bluetooth Chipset.
Current only NXP IW612 Chipset (BT_NXP_NW612) has
been supported.

Add modules/hal_nxp/bt_controller/CMakeLists.txt to
determine whether any firmware is selected, and
check whether the firmware exists.

If the firmware exists, copy the firmware to the
temporary folder ${ZEPHYR_BINARY_DIR}/include/
generated/bt_nxp_ctlr_fw.h. OR, raise a fatal error.

In file hci_nxp_setup.c, includes the temporary file
bt_nxp_ctlr_fw.h.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -238,3 +238,31 @@ config BT_SILABS_HCI_BUFFER_MEMORY
 	  Select the size of allocated memory buffer for the Silicon Labs
 	  Bluetooth Library.
+
+config BT_H4_NXP_CTLR
+	bool "NXP Bluetooth Controller"
```
Wouldn't it make sense to add the following here:
```
default y
depends on DT_HAS_NXP_BT_HCI_UART_ENABLED
```

Then you can remove any explicit enablement in the app Kconfig overlays.

jhedberg replied:
...and then it'll also be easier for me to do the conversion as part of #72323

lylezhu2012 replied:
> Wouldn't it make sense to add the following here:
> 
> ```
> default y
> depends on DT_HAS_NXP_BT_HCI_UART_ENABLED
> ```
> 
> Then you can remove any explicit enablement in the app Kconfig overlays.

Sorry, the dependency cannot be added.

The Bluetooth controller is module, it can work with any MCU with UART peripheral.

jhedberg replied:
@lylezhu2012 I'm not sure I follow. Even in the case that you describe, those MCUs that have this peripheral connected will still need to have the appropriate devicetree node with `compatible = "nxp,bt-hci-uart";`, won't they? If not, what is the purpose of the devicetree node?

lylezhu2012 replied:
`compatible = "nxp,bt-hci-uart";` is the devicetree node required by NXP Bluetooth Controller. If you want to use NXP Bluetooth Controller, you need to add the devicetree node with `compatible = "nxp,bt-hci-uart";`.

jhedberg replied:
@lylezhu2012 the compatible strings are used to associate devicetree nodes with drivers. Which driver implements this compatible? You're adding the `hci_nxp_setup.c` extension driver in this same PR where you add the devicetree binding and compatible, so it seems natural to assume that these go together, even though the c-file currently doesn't `#define DT_DRV_COMPAT nxp_bt_hci_uart` (which I assumed was primarily because it only supports a single instance at the moment).

lylezhu2012 replied:
Yes, `nxp,bt-hci-uart` is used for driver hci_nxp_setup.c.

> @lylezhu2012 the compatible strings are used to associate devicetree nodes with drivers. Which driver implements this compatible? You're adding the `hci_nxp_setup.c` extension driver in this same PR where you add the devicetree binding and compatible, so it seems natural to assume that these go together, even though the c-file currently doesn't `#define DT_DRV_COMPAT nxp_bt_hci_uart` (which I assumed was primarily because it only supports a single instance at the moment).

Yes. Currently, only one instance can be support. 

And I have a question, if does Bluetooth Host support multiple instance? I mean can we call `bt_enable` multiple times to enable multiple Bluetooth Host?

jhedberg replied:
> Yes, `nxp,bt-hci-uart` is used for driver hci_nxp_setup.c.

So in that case you should be able to apply my proposal wrt `default` and `depends on`. If there isn't a node with that compatible string it doesn't make sense to enable the driver.

> And I have a question, if does Bluetooth Host support multiple instance? I mean can we call `bt_enable` multiple times to enable multiple Bluetooth Host?

Not right now. This may become possible at some point in the future, especially thanks to #72323. With that PR you could quite easily construct a host that supports multiple controllers, or an app which exposes multiple controllers over different HCI transports.

lylezhu2012 commented:
Fix conflict issue.

jori-nordic commented:
@lylezhu2012 could you make it also work on **MIMXRT1040-EVK**, as that's the one you suggested (in https://github.com/zephyrproject-rtos/zephyr/issues/67145) and we bought.

lylezhu2012 commented:
> @lylezhu2012 could you make it also work on **MIMXRT1040-EVK**, as that's the one you suggested (in #67145) and we bought.

There is a pin conflict in MIMXRT1040-EVK. If PCM pins for handsfree (or AG) are configured by soldering 0 ohm resistors, the board cannot be flashed. So for BR cases, I do not recommend using MIMXRT1040-EVK. MIMXRT1170-EVKB does not have this limitation. That's why I made this PR to support MIMXRT1170-EVKB.

jori-nordic commented:
> If PCM pins for handsfree (or AG) are configured by soldering 0 ohm resistors

We don't really care about the PCM data, it's only for testing that the ACL links and L2CAP channel still work. We could still test, e.g. RFCOMM or A2DP with all-zeroes PCM data.

That board can also be useful to test LE. 

> the board cannot be flashed

Can you share where in the official documentation I can find this information?

lylezhu2012 commented:
> We don't really care about the PCM data, it's only for testing that the ACL links and L2CAP channel still work. We could still test, e.g. RFCOMM or A2DP with all-zeroes PCM data.
> That board can also be useful to test LE.

Got it. I can enable it after the PR merged.

> Can you share where in the official documentation I can find this information?

No. We flashed it internally by adding an extra circuit. Or, remove the module manually.

lylezhu2012 commented:
> > We don't really care about the PCM data, it's only for testing that the ACL links and L2CAP channel still work. We could still test, e.g. RFCOMM or A2DP with all-zeroes PCM data.
> > That board can also be useful to test LE.
> 
> Got it. I can enable it after the PR merged.
> 
> > Can you share where in the official documentation I can find this information?
> 
> No. We flashed it internally by adding an extra circuit. Or, remove the module or a jumper manually.

lylezhu2012 committed:
driver: bluetooth: hci: Add NXP BT module support

Implement UART firmware download driver for NXP
BT module.

Only support Murata 2EL M.2 module on RT1170EVKB.

And only one instance can be supported now.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
@jori-nordic , I saw that [72323 ](https://github.com/zephyrproject-rtos/zephyr/pull/72323)has been merged. And I updated the PR to support it.

Please help review it.

jori-nordic commented:
Will do. I don't have the HW though so I can only try building :+1:

lylezhu2012 commented:
> Will do. I don't have the HW though so I can only try building 👍

Sure. Please. I have tested it locally.
